### ISSUE TITLE
Large files on Google Docs are corrupted when used by CoLab

### ISSUE THREAD

**Issue by palmcorp** [googlecolab/colabtools#679]
Bug report for Colab: http://colab.research.google.com/.

FWe are working with pre-trained models for a TensorFlow application running on CoLab.  There are 5 files of 1.5+ Gig each.  These pre-trained models are not modified by the Tensor Flow application.  Only a ‘restore’ is done.  The application can run to completion one night and the next night some of the checksums don’t match and the models have to be reloaded.  After reloading the application runs fine.
Is there a way to impose read only lock on the files or force a resync to the CoLab notebook in the event that the just loaded files are in some cached limbo?

**Comment by colaboratory-team** [comment]
You can use `drive.flush_and_unmount()` to synchronously wait for any cached/dirty pages in the Colab VM to be written back to Google Drive.
If that doesn't unblock you please share a minimal repro notebook.

**Comment by palmcorp** [comment]
Thanks for the input. Two points not made clear.  The pre-trained models (~1.5 gig each) are mapped from Google Docs to the CoLab notebook using:   from google.colab import drive; drive.mount ('/content/gdrive'). The TensorFlow code does a restore from the /content/gdrive/ directory.  If it finds a checkpoint problem TF stops. Then the models are saved again to the Google Docs repository from masters.  Usually this works and the TF can be restarted.  So the issue is probably the other direction, From Google Docs Drive to CoLab VM.  What is puzzling is that the pre-trained model files are never written to from CoLab since this is a transfer learning exercise.  I see three possibilities: 1) Somehow the Google Docs files are being corrupted [unlikely],  2) When CoLab reads the large files from Google Docs there is some communication corruption [hard to believe] 3) The files show up in the CoLab but aren’t really completely there [?].  Any ideas?

**Comment by colaboratory-team** [comment]
Thanks for clarifying.  1.5GiB is far too large to expect writes to be atomic, so it's likely that you're observing "tearing": the TF reader in colab ends up reading bytes from two different generations of the file (i.e. across "save" operations; still unclear on what process is doing the writing and how it's doing it; is it another colab VM using a drive.mount()'ed view, or is it an unrelated process on using the Google Drive API, or something else?).
It might be helpful to make the data immutable: have you writer write to a new location (e.g. under a folder whose name contains a globally-unique timestamp), and once done writing, update a pointer to that location. Then your reader can read the pointer to know where to read from, and be assured that the read source won't change out from under it.

**Comment by palmcorp** [comment]
Appreciate the feedback.  Actually no processes are writing to these models.  They are ‘pre-trained’ TF models.  The program under test restores the session from the pre-trained model and proceeds from there.  In ‘transfer learning’ the goal is to take advantage of models that have been trained on 1,000’s of tests.  As mentioned above the TF process under study does not conduct any ‘save.’  It just restores , i.e. saver.restore (..), so the models should not be changed.  This is the puzzling part and also that sometimes they are used w/o problem and then a checksum doesn’t match.  Would it be safer to import the models into the CoLab VM before using?  Might take an hour to load but maybe the only answer.  Comments?

**Comment by palmcorp** [comment]
Question: is this a TensorFlow problem or a CoLab<->Google Docs issue? Any ideas where to look?  Checked the data flow of 1.5 Gig file from Google Docs to CoLab Vm transfer and it goes quickly (< 2 min) so can't be an issue with time out.  What data from our side would help resolve this issue?
PS please don't close the ticket.

**Comment by colaboratory-team** [comment]
I'm confused; earlier you said "the models are saved again to the Google Docs repository from masters" but then you said "no processes are writing to these models". Are the files ever changed?
If no, then I'd pre-copy them from Drive to the VM, which will make everything more robust and more efficient (assuming most of the data is eventually read by the VM anyway).
If yes then I need to understand more about your over-all system layout (what's writing, what's reading, etc).

**Comment by palmcorp** [comment]
Ok. Here is the answer.  As you commented and I suspected, there is some issue with accessing large files on the Google Docs account from CoLab.  Maybe some timeout, possibly communication error, etc.  Now we keep the pre-trained model  archive on Google Docs.  Before running our application the models are copied to CoLab VM using (!cp ...).  The copying is relatively fast even though we are talking about ~ 10Gig.  The program runs in the CoLab VM.  Since using this strategy there have been no problems.  I'm writing this so others may benefit and Google can investigate the root cause.  (Contact me if additional details are needed.)

**Comment by colaboratory-team** [comment]
@palmcorp thanks for the confirmation and offer for additional details. If you can share a minimal repro notebook with colaboratory-team@google.com that'd be fantastic.

**Comment by cibic89** [comment]
Experiencing a similar issue. I cannot extract large archives in google colab from my google drive if the extract directory is within google drive... I tested the archive numerous times and no errors occur. This only happens when I access the archive from google drive...

**Comment by palmcorp** [comment]
Solved by using !cp  within colab to copy from Google Docs to Colab.  This moves the files into the colab virtual space.  Then use them directly.  Requires a copy but worth the effort.   I have 14 Gig of pre-trained models and it takes 10-15 minutes to copy over, but the copy subprogram remains in the Juniper book on colab  - so reuse if possible .

 

-- 

==============================

Paul L Petronelli

CEO

PALM Associates, Inc.

palm@palmcorp.com

 

(408) 254-8200 [office]

(408) 582-8056 [cell]

http://www.palmcorp.com

 

These statements are subject to the PALM Associates, Inc.  Proprietary &

Confidentiality Agreement a.k.a. NDA. This is not an offer, even if terms

of an offer are discussed hypothetically. Pursuant to the express terms of

the NDA, only offers for sale that are the subject of a written definitive

agreement, signed by all the parties to that agreement, shall have any

legal effect. Prior to such definitive agreements, if any, no discussion

between and among the parties shall have any legal effect whatsoever.
